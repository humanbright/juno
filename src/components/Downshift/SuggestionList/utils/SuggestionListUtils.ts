import { RcClasses } from '../../../../foundation';
import { InnerSuggestionListProps } from '../SuggestionList';

export const RcSuggestionListClasses = RcClasses<InnerSuggestionListProps>(
  ['root', 'expanded', 'toggle'],
  'RcSuggestionList',
);
