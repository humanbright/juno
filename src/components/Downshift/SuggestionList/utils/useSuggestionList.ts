import { ChangeEvent, HTMLAttributes, useMemo, useRef } from 'react';

import uniqueId from 'lodash/uniqueId';

import { useControlled } from '@material-ui/core/utils';

import {
  combineProps,
  useForceUpdate,
  useKeyboardMoveFocus,
  useResultRef,
} from '../../../../foundation';
import { RcIconButtonProps } from '../../../Buttons/IconButton';
import { RcTextFieldProps } from '../../../Forms';
import { RcDownshiftProps } from '../../Downshift';
import {
  DEFAULT_GET_OPTION_LABEL,
  isItemCanSelected,
  RcDownshiftGetItemPropsOptions,
  RcDownshiftHighlightChangeReason,
  RcDownshiftSelectedItem,
} from '../../utils';
import { useDownshiftGroup } from '../../utils/useDownshiftGroup';

const DEFAULT_HIGHLIGHTED_INDEX = -1;

type UseDownshiftParams<T = RcDownshiftSelectedItem> = {
  /** that input ref you binding event */
  inputRef: React.RefObject<HTMLInputElement>;
  /** emit when item be select */
  onSelect?: (event: React.ChangeEvent<{}>, selectedItems: T) => void;
} & Pick<
  RcDownshiftProps<T>,
  | 'options'
  | 'value'
  | 'getOptionLabel'
  | 'filterOptions'
  | 'inputValue'
  | 'onInputChange'
  | 'onKeyDown'
  | 'onClear'
  | 'getOptionDisabled'
  | 'disabledItemsHighlightable'
  | 'variant'
  | 'groupBy'
  | 'getExpandIconProps'
  | 'groupExpanded'
  | 'groupDefaultExpanded'
  | 'onGroupExpanded'
  | 'groupVariant'
>;

const componentName = 'useSuggestionList';

/**
 * provide suggestion list with search function and expandable group with virtualized list
 */
export const useSuggestionList = <
  T extends RcDownshiftSelectedItem = RcDownshiftSelectedItem,
>({
  inputValue: inputValueProp,
  getOptionLabel = DEFAULT_GET_OPTION_LABEL,
  filterOptions,
  disabledItemsHighlightable,
  options,
  onInputChange: onInputChangeProp,
  inputRef,
  onKeyDown: onKeyDownProp,
  onSelect,
  onClear,
  getOptionDisabled,
  groupBy,
  onGroupExpanded,
  groupVariant,
  groupExpanded,
  groupDefaultExpanded,
  getExpandIconProps,
}: UseDownshiftParams<T>) => {
  const [inputValue, setInputValue] = useControlled({
    controlled: inputValueProp,
    default: '',
    name: componentName,
  });

  const updateInputValue = (newValue: string) => {
    onInputChangeProp?.(newValue);
    setInputValue(newValue);
  };

  const focusInput = () => inputRef.current?.focus();

  const { current: suggestionListId } = useResultRef(() =>
    uniqueId(`suggestion-list-`),
  );
  // * use -2 for us know that is init state, for recalculate defaultHighlightedIndex
  const highlightedIndexRef = useRef(DEFAULT_HIGHLIGHTED_INDEX);
  const changeHighlightedIndexReason =
    useRef<RcDownshiftHighlightChangeReason>();

  const forceUpdate = useForceUpdate();

  const filteredResult = useMemo(() => {
    const getFilteredItems = (items: T[]) => {
      if (filterOptions) {
        return filterOptions(items, {
          inputValue,
          getOptionLabel,
          selectedItems: [],
        });
      }

      return items;
    };

    // * only when isOpen calculate the filtered result
    const results = getFilteredItems(options!);

    return results;
  }, [filterOptions, getOptionLabel, inputValue, options]);

  const { groupedResult, handleGroupExpandedChange, optionsGroupList } =
    useDownshiftGroup<T>({
      groupBy,
      options,
      filteredResult,
      getExpandIconProps,
      groupExpanded,
      groupDefaultExpanded,
      onGroupExpanded,
      groupVariant,
      getOptionDisabled,
      id: suggestionListId,
    });

  const optionItems = groupBy ? groupedResult : filteredResult;

  const setHighlightedIndex = (
    index: number,
    {
      reRender = false,
      reason,
    }: {
      reRender?: boolean;
      reason?: RcDownshiftHighlightChangeReason;
    },
  ) => {
    changeHighlightedIndexReason.current = reason;

    if (highlightedIndexRef.current !== index) {
      highlightedIndexRef.current = index;

      if (reRender) forceUpdate();
    }
  };

  const handleInputChange = (newValue: string) => {
    if (inputValue !== newValue) {
      setHighlightedIndex(DEFAULT_HIGHLIGHTED_INDEX, { reason: 'auto' });

      updateInputValue(newValue);
    }

    return false;
  };

  const getIsItemCanSelected = (item?: T | null): item is T => {
    return (
      (!!item && item.freeSolo) ||
      (isItemCanSelected(item) && !getOptionDisabled?.(item))
    );
  };

  const selectItemFn = (e: ChangeEvent<{}>, selectedItem: T | null) => {
    if (getIsItemCanSelected(selectedItem)) {
      onSelect?.(e, selectedItem);

      return true;
    }
    return false;
  };

  const resetState = () => {
    if (inputRef.current && inputRef.current.value.length > 0) {
      updateInputValue('');
    }
  };

  const reset = (isFocus?: boolean) => {
    resetState();

    onInputChangeProp?.('');
    if (isFocus) focusInput();
  };

  const { onKeyFocusedIndexHandle } = useKeyboardMoveFocus<T>({
    options: optionItems,
    focusedIndexRef: highlightedIndexRef,
    infinite: true,
    onFocusedIndexChange: (event, toIndex) => {
      setHighlightedIndex(toIndex, { reason: 'keyboard', reRender: true });

      event?.preventDefault();
    },
    getOptionDisabled: disabledItemsHighlightable
      ? undefined
      : (child) => {
          return !getIsItemCanSelected(child);
        },
  });

  const getItemProps = ({
    item,
    index,
    ...itemRest
  }: RcDownshiftGetItemPropsOptions<T>) => {
    return combineProps(
      {
        id: `${suggestionListId}-option-${index}`,
        role: 'option',
        onClick: (e) => {
          selectItemFn(e, item);
        },
        onMouseDown: (e) => {
          e.preventDefault();
          e.stopPropagation();
        },
        onMouseOver: () => {
          if (highlightedIndexRef.current !== index) {
            setHighlightedIndex(index!, { reason: 'mouse', reRender: true });
          }
        },
      },
      itemRest,
    );
  };

  const getInputProps = (props?: RcTextFieldProps['InputProps']) => {
    return combineProps(
      {
        id: `${suggestionListId}-input`,
        autoComplete: 'off',
        onChange: (e) => {
          const changeValue = e.target.value;
          handleInputChange(changeValue);
        },
        onKeyDown: (e) => {
          onKeyDownProp?.(e, highlightedIndexRef.current);

          const highlightedIndex = highlightedIndexRef.current;
          switch (e.key) {
            case 'F10':
              if (e.shiftKey) {
                const currOption = optionItems[highlightedIndex];
                const currentGroup = currOption.group;

                if (currentGroup && currentGroup.options.length > 1) {
                  handleGroupExpandedChange(currentGroup.group);
                }
              }
              break;
            case 'Enter':
              if (e.which === 229) return;
              if (highlightedIndex !== DEFAULT_HIGHLIGHTED_INDEX) {
                const currOption = optionItems[highlightedIndex];
                const isGroupTitle =
                  currOption === currOption.group?.options[0];

                if (isGroupTitle) {
                  const { onClick } = currOption.group!.getExpandIconProps!()!;

                  onClick?.(e as any);
                } else {
                  selectItemFn(e, currOption);
                }

                e.stopPropagation();
              }
              // always preventDefault for not inset all enter into textarea
              e.preventDefault();
              break;
            default:
              onKeyFocusedIndexHandle(e);
              break;
          }
        },
      },
      props,
    );
  };

  const getInputAriaProps = (props?: RcTextFieldProps['inputProps']) => {
    return combineProps(
      {
        // * provide for when container click focus on input
        onContainerClick: focusInput,
        role: 'combobox',
        'aria-autocomplete': 'list',
        // TODO
        'aria-expanded': true,
        'aria-haspopup': true,
        'aria-owns': `${suggestionListId}-menu`,
        'aria-activedescendant':
          highlightedIndexRef.current > -1
            ? `${suggestionListId}-option-${highlightedIndexRef.current}`
            : undefined,
      },
      props,
    );
  };

  const getLabelProps = (props?: RcTextFieldProps['InputLabelProps']) => {
    return combineProps(
      {
        htmlFor: `${suggestionListId}-input`,
        id: `${suggestionListId}-label`,
      },
      props,
    );
  };

  const getMenuProps = (restMenuProps?: HTMLAttributes<HTMLElement>) => {
    return combineProps(
      {
        'aria-labelledby': `${suggestionListId}-label`,
        id: `${suggestionListId}-menu`,
        role: 'listbox',
      },
      restMenuProps,
    );
  };

  const getClearButtonProps = (props?: RcIconButtonProps) => {
    return combineProps(
      {
        id: `${suggestionListId}-clear-button`,
        onClick: (e) => {
          onClear?.(e);
          reset(true);
        },
      },
      props,
    );
  };

  const resultObj = {
    focusInput,
    getClearButtonProps,
    getLabelProps,
    getMenuProps,
    getInputProps,
    getInputAriaProps,
    getItemProps,
    highlightedIndex: highlightedIndexRef.current,
    optionItems,
    inputValue,
    onInputChange: handleInputChange,
    setHighlightedIndex,
    changeHighlightedIndexReason: changeHighlightedIndexReason.current,
    reset,
    forceUpdate,
    optionsGroupList,
  };

  changeHighlightedIndexReason.current = undefined;
  return resultObj;
};
