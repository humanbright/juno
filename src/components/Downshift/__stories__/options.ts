import range from 'lodash/range';

import { RcDownshiftSelectedItem } from '../utils';

export const options = [
  { id: 'phone_1', label: 'Afghanistan' },
  { id: 'phone_2', label: 'Aland Islands' },
  { id: 'phone_3', label: 'Bahamas' },
  { id: 'phone_4', label: 'Bahrain' },
  { id: 'phone_5', label: 'China' },
  { id: 'phone_6', label: 'Algeria' },
  { id: 'phone_7', label: 'American Samoa' },
  { id: 'phone_8', label: 'Andorra' },
  { id: 'phone_9', label: 'Angola' },
  { id: 'phone_10', label: 'Anguilla' },
  { id: 'phone_11', label: 'Antarctica' },
  { id: 'phone_12', label: 'Antigua and Barbuda' },
  { id: 'phone_13', label: 'Argentina' },
  { id: 'phone_14', label: 'Armenia' },
  {
    id: 'phone_15',
    label: "This item couldn't Selectable",
    unSelectable: true,
  },
  { id: 'phone_16', label: '木星' },
  { id: 'phone_16_2', label: '木星_1' },
  { id: 'phone_16_3', label: '木星_2' },
  { id: 'phone_16_4', label: '木星_3' },
  { id: 'phone_17', label: '鈴盛軟件' },
  { id: 'phone_18', label: '整合' },
  ...range(19, 10000).map((id: number) => ({
    id: `phone_${id}`,
    label: `Item-${id}`,
  })),
].sort((a, b) => {
  const nameA = a.label.toUpperCase(); // ignore upper and lowercase
  const nameB = b.label.toUpperCase(); // ignore upper and lowercase
  if (nameA < nameB) {
    return -1;
  }
  if (nameA > nameB) {
    return 1;
  }

  // names must be equal
  return 0;
}) as RcDownshiftSelectedItem[];
