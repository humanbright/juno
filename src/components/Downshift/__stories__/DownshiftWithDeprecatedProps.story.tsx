import { boolean, text } from '@storybook/addon-knobs';
import { Story } from '@storybook/react/types-6-0';
import differenceBy from 'lodash/differenceBy';
import React, { ComponentProps, forwardRef, useState } from 'react';

import { RcChip } from '../../Chip';
import { RcMenuItem } from '../../Menu';
import { RcTypography } from '../../Typography';
import { RcDownshift } from '../Downshift';
import { RcDownshiftSelectedItem } from '../utils';
import { options } from './options';

export default {
  title: 'ðŸ–¤ Deprecated Components/Downshift (New with deprecated props)',
  component: RcDownshift,
};

const getKnobs = () => {
  const multiple = boolean('multiple', true);
  const autoSwitchEmail = boolean('autoSwitchEmail', false);
  const enableFreeChips = boolean('enableFreeChips', true);
  const allowPlainHelperText = boolean('allowPlainHelperText', false);
  const enableAutoTransform = boolean('enableAutoTransform', false);
  const initialIsOpen = boolean('initialIsOpen', false);
  const defaultIsOpen = boolean('defaultIsOpen', false);

  const disabled = boolean('disabled', false);

  return {
    multiple,
    disabled,
    autoSwitchEmail,
    enableFreeChips,
    allowPlainHelperText,
    enableAutoTransform,
    initialIsOpen,
    defaultIsOpen,
  };
};

const suggestions = options;

const getTextFieldKnobs = () => {
  const required = boolean('required', false);
  const placeholder = text('placeholder', 'Placeholder');
  const helperText = text('helperText', 'helperText');
  const autoFocus = boolean('autoFocus', false);
  const inputAriaLabel = text('input aria-label', 'input aria label content');
  const label = text('label', 'Downshift');

  return {
    label,
    placeholder,
    helperText,
    autoFocus,
    required,
    inputAriaLabel,
  };
};

const getItemById = (suggestions: RcDownshiftSelectedItem[], id: number) =>
  suggestions.find(
    (suggestion: RcDownshiftSelectedItem) => suggestion.id === id,
  );

const SearchItem = forwardRef((props: any, ref) => {
  const { itemId, isHighlighted, label, ...rest } = props;
  const item = getItemById(suggestions, itemId);

  return item ? (
    <RcMenuItem ref={ref} selected={isHighlighted} {...rest} component="div">
      {label}
    </RcMenuItem>
  ) : null;
});

const Chip = forwardRef((props: any, ref) => {
  const item = getItemById(suggestions, props.id);

  return item ? (
    <RcChip
      ref={ref}
      {...props}
      deleteTooltip="Delete"
      data-is-suggestion={item.isSuggestion}
    />
  ) : (
    <RcChip
      {...props}
      ref={ref}
      deleteTooltip="Delete"
      data-is-suggestion={props.isSuggestion}
    />
  );
});

export const Downshift: Story<ComponentProps<typeof RcDownshift>> = () => {
  const knobsProps = getKnobs();
  const { inputAriaLabel, label, ...textFieldRest } = getTextFieldKnobs();

  const [inputValue, setInputValue] = useState('');

  const [suggestionItems, setSuggestionItems] = useState<
    RcDownshiftSelectedItem[]
  >(
    // Provides initial data when `initialIsOpen` is `true`,
    // otherwise we can not see the effect of `initialIsOpen`
    knobsProps.initialIsOpen ? suggestions : [],
  );
  const [selectedItems, setSelectedItems] = useState<RcDownshiftSelectedItem[]>(
    [
      { id: 'phone_1', label: 'Afghanistan' },
      { id: 'phone_2', label: 'Aland Islands' },
    ],
  );

  const handleInputChange = (value: string) => {
    console.log('InputChange', value);
    const inputValue = value;
    setInputValue(inputValue);

    const inputLength = inputValue.length;
    let filterSuggestions: any = [];

    if (inputLength === 0) {
      return setSuggestionItems([]);
    }
    filterSuggestions = differenceBy(suggestions, selectedItems, 'id');
    filterSuggestions = filterSuggestions.filter((suggestion: any) => {
      const keep =
        suggestion.label.slice(0, inputLength).toLowerCase() === inputValue;

      return keep;
    });
    console.log(filterSuggestions);
    setSuggestionItems(filterSuggestions);
  };
  const handleSelectChange = (selectedItems: RcDownshiftSelectedItem[]) => {
    console.log(selectedItems);
    setSelectedItems(selectedItems);
    setInputValue('');
  };

  return (
    <div>
      <RcTypography color="highlight.f02">
        Type 'a' in the text field, that will be dropdown menu show to select
      </RcTypography>
      <RcTypography color="danger.f02">
        Don't click suggestion items before composition end if you use microsoft
        pinyin
      </RcTypography>
      <br />
      <RcDownshift
        {...knobsProps}
        suggestionItems={suggestionItems}
        MenuItem={SearchItem}
        label={label}
        gutterBottom
        InputItem={Chip}
        onInputChange={handleInputChange}
        onSelectChange={handleSelectChange}
        minRowHeight={50}
        inputValue={inputValue}
        screenReader={{
          entry: '1 entry',
          entries: `${selectedItems.length} entries`,
        }}
        selectedItems={selectedItems}
        TextFieldProps={{
          inputProps: {
            'aria-label': inputAriaLabel,
          },
          ...textFieldRest,
        }}
      />
    </div>
  );
};

Downshift.args = {};

Downshift.storyName = 'Downshift (New with deprecated props)';
