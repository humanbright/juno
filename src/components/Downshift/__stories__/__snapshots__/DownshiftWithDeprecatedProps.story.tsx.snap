// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Storyshots ðŸ–¤ Deprecated Components/Downshift (New with deprecated props) Downshift (New with deprecated props) 1`] = `
.c11 {
  width: 1em;
  height: 1em;
  stroke-width: 0;
  stroke: currentColor;
  fill: currentColor;
  pointer-events: none;
}

.c10 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  font-size: 24px;
}

.c8.RcIconButton-root {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: 20px;
  height: 20px;
  color: #666666;
  border-radius: 50%;
  -webkit-transition: background 150ms cubic-bezier(0.4,0,0.2,1) 0ms;
  transition: background 150ms cubic-bezier(0.4,0,0.2,1) 0ms;
  cursor: pointer;
}

.c8.RcIconButton-root .c9 {
  font-size: 20px;
}

.c8.RcIconButton-root:active .c9 {
  color: #666666;
}

.c8.RcIconButton-root.focus-visible,
.c8.RcIconButton-root[data-focus-visible-added] {
  background-color: transparent;
}

.c8.RcIconButton-root.focus-visible:active .c9,
.c8.RcIconButton-root[data-focus-visible-added]:active .c9 {
  color: rgba(102,102,102,0.88);
}

.c8.RcIconButton-root.focus-visible:after,
.c8.RcIconButton-root[data-focus-visible-added]:after {
  content: '';
  position: absolute;
  top: -4px;
  right: -4px;
  bottom: -4px;
  left: -4px;
  box-shadow: 0 0 0 1px #066FAC;
  border-radius: 50%;
  pointer-events: none;
  z-index: 1;
}

.c8.RcIconButton-persistBg {
  background-color: rgba(102,102,102,0.12);
}

.c8.RcIconButton-disabled .c9 {
  color: #B9B9B9;
}

.c8.RcIconButton-invisible {
  opacity: 0;
  visibility: hidden;
  width: 0;
  height: 0;
}

.c8.RcIconButton-outline {
  border: 1px solid #C7C7C7;
}

.c8.RcIconButton-contained {
  color: #fff;
  background-color: #666666;
}

.c8.RcIconButton-contained:active {
  background-color: rgb(91,91,91);
}

.c8.RcIconButton-contained:active .c9 {
  color: #fff;
}

.c8.RcIconButton-contained.focus-visible,
.c8.RcIconButton-contained[data-focus-visible-added] {
  background-color: rgb(91,91,91);
}

.c8.RcIconButton-contained.focus-visible:active .c9,
.c8.RcIconButton-contained[data-focus-visible-added]:active .c9 {
  color: #fff;
}

.c8 .RcIconButtonTouchRipple-ripplePulsate {
  border-radius: 0;
  -webkit-animation-name: none;
  animation-name: none;
  opacity: 0;
}

.c6 {
  height: 28px;
  box-sizing: border-box;
  padding: 2px;
  background-color: #F3F3F3;
  color: #121212;
  position: relative;
  border-radius: 100vw;
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
  font-size: 0.9375rem;
  font-weight: 400;
  font-family: Lato,Helvetica,Arial,sans-serif;
  line-height: 22px;
}

.c6.focus-visible,
.c6[data-focus-visible-added] {
  background-color: rgba(73,73,73,0.08);
}

.c6.focus-visible:after,
.c6[data-focus-visible-added]:after {
  content: '';
  position: absolute;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  box-shadow: inset 0 0 0 1px #066FAC;
  border-radius: 100vw;
  pointer-events: none;
}

.c6:focus {
  background-color: rgba(73,73,73,0.16);
}

.c6:active {
  background-color: rgba(73,73,73,0.24);
}

.c6.RcChip-disabled {
  color: #B9B9B9;
  opacity: 1;
}

.c6 .RcChip-avatar {
  color: #FFFFFF;
  margin-left: 0px;
  margin-right: -4px;
}

.c6 .RcChip-avatar span {
  width: auto;
  height: auto;
}

.c12 {
  -webkit-clip-path: inset(100%);
  clip-path: inset(100%);
  -webkit-clip: rect(1px,1px,1px,1px);
  clip: rect(1px,1px,1px,1px);
  height: 1px;
  overflow: hidden;
  position: absolute;
  white-space: nowrap;
  width: 1px;
}

.c2 .c7 {
  margin-left: 12px;
}

.c2.RcTextField-root {
  margin: 0 0 16px 0;
}

.c2 .RcTextFieldInputLabel-root {
  font-size: 0.75rem;
  font-weight: 700;
  font-family: Lato,Helvetica,Arial,sans-serif;
  line-height: 16px;
  color: #666666;
  -webkit-transform: scale(1);
  -ms-transform: scale(1);
  transform: scale(1);
}

.c2 .RcTextFieldInputLabel-focused {
  color: #066FAC;
}

.c2 .RcTextFieldInput-root {
  color: #121212;
}

.c2 .RcTextFieldInput-root .sc-carFqZ {
  display: none;
}

.c2 .RcTextFieldInput-input {
  font-size: 1rem;
  font-weight: 400;
  font-family: Lato,Helvetica,Arial,sans-serif;
  line-height: 24px;
}

.c2 .RcTextFieldInput-input::-webkit-input-placeholder {
  color: #757575;
  opacity: 1;
}

.c2 .RcTextFieldInput-input::-moz-placeholder {
  color: #757575;
  opacity: 1;
}

.c2 .RcTextFieldInput-input:-ms-input-placeholder {
  color: #757575;
  opacity: 1;
}

.c2 .RcTextFieldInput-input::placeholder {
  color: #757575;
  opacity: 1;
}

.c2 .RcTextFieldInput-underline:before {
  border-bottom-color: #939393;
}

.c2 .RcTextFieldInput-underline:after {
  border-bottom-color: #066FAC;
}

.c2 .RcTextFieldInput-error:after {
  border-bottom-color: #BE3933;
}

.c2 .RcTextFieldInput-disabled {
  color: #B9B9B9;
  -webkit-text-fill-color: #B9B9B9;
}

.c2 .RcTextFieldInput-disabled::-webkit-input-placeholder {
  color: #B9B9B9;
}

.c2 .RcTextFieldInput-disabled::-moz-placeholder {
  color: #B9B9B9;
}

.c2 .RcTextFieldInput-disabled:-ms-input-placeholder {
  color: #B9B9B9;
}

.c2 .RcTextFieldInput-disabled::placeholder {
  color: #B9B9B9;
}

.c2 .RcTextFieldInput-disabled:before {
  border-bottom-style: solid;
  border-bottom-color: #B9B9B9;
}

.c2 .RcTextFieldInput-disabled:after {
  border-bottom-style: none;
}

.c2 .RcTextFieldFormHelperText-root {
  font-size: 0.75rem;
  font-weight: 400;
  font-family: Lato,Helvetica,Arial,sans-serif;
  line-height: 16px;
  color: #666666;
}

.c2 .RcTextFieldFormHelperText-error {
  color: #BE3933;
}

.c2 .RcTextFieldFormHelperText-disabled {
  color: #B9B9B9;
}

.c2 .RcTextFieldInputLabel-error {
  color: #BE3933;
}

.c2 .RcTextFieldInputLabel-disabled {
  color: #B9B9B9;
}

.c2 input[type='search']::-ms-clear,
.c2 input[type='search']::-ms-reveal {
  display: none;
  width: 0;
  height: 0;
}

.c2 input[type='search']::-webkit-search-decoration,
.c2 input[type='search']::-webkit-search-cancel-button,
.c2 input[type='search']::-webkit-search-results-button,
.c2 input[type='search']::-webkit-search-results-decoration {
  display: none;
}

.c0 {
  font-size: 0.9375rem;
  font-weight: 400;
  font-family: Lato,Helvetica,Arial,sans-serif;
  line-height: 22px;
  color: #FF8800;
}

.c1 {
  font-size: 0.9375rem;
  font-weight: 400;
  font-family: Lato,Helvetica,Arial,sans-serif;
  line-height: 22px;
  color: #BE3933;
}

.c3 .RcDownshiftInput-input {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  min-width: 30px;
}

.c3 .RcTextFieldInput-focused .sc-carFqZ {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
}

.c4 .RcDownshiftInput-root {
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  padding: 0 0px 0 0;
}

.c4 .RcDownshiftInput-root .c5 {
  margin: 2px;
}

.c4 .RcDownshiftInput-root .c5 .RcChip-deleteIcon {
  margin-left: -6px;
}

.c4 .RcDownshiftInput-root:before,
.c4 .RcDownshiftInput-root:after {
  z-index: 1;
}

.c4 .RcDownshiftInput-container {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  overflow-y: auto;
  max-height: 132px;
  width: 100%;
}

@media (hover:hover) and (pointer:fine) {
  .c8.RcIconButton-root:hover {
    background-color: transparent;
  }

  .c8.RcIconButton-root:hover .c9 {
    color: rgba(102,102,102,0.88);
  }
}

@media (hover:hover) and (pointer:fine) {
  .c8.RcIconButton-contained:hover {
    background-color: rgb(91,91,91);
  }

  .c8.RcIconButton-contained:hover .c9 {
    color: #fff;
  }
}

@media (hover:hover) and (pointer:fine) {
  .c6:hover {
    background-color: rgba(73,73,73,0.12);
  }
}

@media (hover:hover) and (pointer:fine) {
  .c2 .RcTextFieldInput-underline:hover:not(.RcTextFieldInput-disabled):before {
    border-bottom-color: #121212;
  }
}

@media (hover:hover) and (pointer:fine) {
  .c3 .RcTextFieldInput-root:hover:not(.RcTextFieldInput-disabled) .sc-carFqZ {
    display: -webkit-inline-box;
    display: -webkit-inline-flex;
    display: -ms-inline-flexbox;
    display: inline-flex;
  }
}

<div>
  <p
    className="MuiTypography-root c0"
    data-color="highlight.f02"
    data-variant="body1"
  >
    Type 'a' in the text field, that will be dropdown menu show to select
  </p>
  <p
    className="MuiTypography-root c1"
    data-color="danger.f02"
    data-variant="body1"
  >
    Don't click suggestion items before composition end if you use microsoft pinyin
  </p>
  <br />
  <div
    className="MuiFormControl-root MuiTextField-root RcTextField-root c2 c3 c4 MuiFormControl-fullWidth"
  >
    <label
      className="MuiFormLabel-root MuiInputLabel-root RcTextFieldInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink"
      data-shrink={true}
      htmlFor="downshift-0-input"
      id="downshift-0-label"
    >
      Downshift
    </label>
    <div
      aria-disabled={false}
      aria-label="Downshift"
      aria-required={false}
      className="MuiInputBase-root MuiInput-root RcTextFieldInput-root RcDownshiftInput-root MuiInput-underline RcTextFieldInput-underline MuiInputBase-fullWidth MuiInput-fullWidth MuiInputBase-formControl MuiInput-formControl"
      onClick={[Function]}
      onCompositionEnd={[Function]}
      onCompositionStart={[Function]}
      onPaste={[Function]}
    >
      <div
        className="RcDownshiftInput-container"
        onClick={[Function]}
      >
        <div
          className="MuiChip-root c5 c6 MuiChip-deletable"
          data-item-last={false}
          data-tag-index={0}
          data-test-automation-class="selected-item"
          data-test-automation-value="phone_1"
          id="phone_1"
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          role="button"
          tabIndex={0}
        >
          <span
            className="MuiChip-label"
          >
            Afghanistan
          </span>
          <button
            aria-describedby={null}
            aria-disabled={false}
            aria-label="Delete"
            className="MuiButtonBase-root c7 c8 MuiChip-deleteIcon RcChip-deleteIcon RcIconButton-root"
            disabled={false}
            onBlur={[Function]}
            onClick={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseOver={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            tabIndex={-1}
            title="Delete"
            type="button"
          >
            <span
              className="c9 c10 delete_circle icon"
            >
              <svg
                className="c11"
                viewBox="0 0 32 32"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M16 2c7.732 0 14 6.268 14 14s-6.268 14-14 14S2 23.732 2 16 8.268 2 16 2zm-4.243 8.343a.999.999 0 00-1.497 1.32l.083.094 4.243 4.241-4.243 4.244a.999.999 0 001.32 1.497l.094-.083L16 17.412l4.243 4.244a.999.999 0 001.497-1.32l-.083-.094-4.243-4.244 4.243-4.241a.999.999 0 00-1.32-1.497l-.094.083L16 14.584l-4.243-4.241z"
                />
              </svg>
              
            </span>
          </button>
        </div>
        <div
          className="MuiChip-root c5 c6 MuiChip-deletable"
          data-item-last={true}
          data-tag-index={1}
          data-test-automation-class="selected-item"
          data-test-automation-value="phone_2"
          id="phone_2"
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          role="button"
          tabIndex={-1}
        >
          <span
            className="MuiChip-label"
          >
            Aland Islands
          </span>
          <button
            aria-describedby={null}
            aria-disabled={false}
            aria-label="Delete"
            className="MuiButtonBase-root c7 c8 MuiChip-deleteIcon RcChip-deleteIcon RcIconButton-root"
            disabled={false}
            onBlur={[Function]}
            onClick={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseOver={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            tabIndex={-1}
            title="Delete"
            type="button"
          >
            <span
              className="c9 c10 delete_circle icon"
            >
              <svg
                className="c11"
                viewBox="0 0 32 32"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M16 2c7.732 0 14 6.268 14 14s-6.268 14-14 14S2 23.732 2 16 8.268 2 16 2zm-4.243 8.343a.999.999 0 00-1.497 1.32l.083.094 4.243 4.241-4.243 4.244a.999.999 0 001.32 1.497l.094-.083L16 17.412l4.243 4.244a.999.999 0 001.497-1.32l-.083-.094-4.243-4.244 4.243-4.241a.999.999 0 00-1.32-1.497l-.094.083L16 14.584l-4.243-4.241z"
                />
              </svg>
              
            </span>
          </button>
        </div>
        <input
          aria-autocomplete="list"
          aria-describedby="downshift-0-input-helper-text"
          aria-expanded={false}
          aria-haspopup={true}
          aria-invalid={false}
          aria-label="input aria label content"
          autoComplete="off"
          autoFocus={false}
          className="MuiInputBase-input MuiInput-input RcTextFieldInput-input RcDownshiftInput-input"
          disabled={false}
          id="downshift-0-input"
          onAnimationStart={[Function]}
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          required={false}
          role="combobox"
          type="text"
          value=""
        />
      </div>
    </div>
  </div>
  <span
    className="c12"
    id="downshift-0-input-helper-text"
  >
    2 entries
  </span>
</div>
`;
